### PR TITLE
Profiler summary

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?
Reaction to #1253. in particular:
- moves profilers do `debugging` package
- created abstraction over Profiler writing streams and add summary methods returning a string

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/debugging/__init__.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
+"""
```
Comment (Borda) [inline pytorch_lightning/debugging/__init__.py]: this is just a full copy-paste

**Comment by Borda** [comment]
@jeremyjordan mind have a look at it before finalizing?

**Comment by jeremyjordan** [comment]
i don't understand the package rename. we went from `logging/` to `loggers/` but now we're changing `profiler/` to `profiling/`? it seems like an unnecessary change.

i like the abstraction over writing the summary more flexibly, however if you're going to rename `describe` to `summary`, then you'll want to update the training loop teardown method to call `summary` instead.

**Comment by Borda** [comment]
> i don't understand the package rename. we went from `logging/` to `loggers/` but now we're changing `profiler/` to `profiling/`? it seems like an unnecessary change.

The case with with `logging` was that there was collision with native python logging package and it was not clear which was imposed as root `__init__` imported both 
In parallel to `loggers` here we also have multiple profilers so plural make more sense, right? 

> i like the abstraction over writing the summary more flexibly, however if you're going to rename `describe` to `summary`, then you'll want to update the training loop teardown method to call `summary` instead.

It is not renaming at all as it may appear from diff, just adding extra function to generate the string text and so there is no no change for user at all (except before you write it file or console, no always write console and optional file)

**Comment by jeremyjordan** [comment]
> It is not renaming at all as it may appear from diff, just adding extra function to generate the string text

ahh i see the `describe` method is just in the base class now and it calls the abstract method `summary`. makes sense, i like this refactor.

however, i'd say that we should keep the module where it's currently located in `pytorch_lightning/profiler/`. changing the location doesn't seem to give us much benefit but then we're left with adding more deprecation warnings and asking users to change their code. as a user it might be frustrating to have to change your import from `pytorch_lightning.profiler` to `pytorch_lightning.profilers`

**Comment by williamFalcon** [comment]
I agree with @jeremyjordan we need to stop making unnecessary changes to package names @Borda

File: pytorch_lightning/profiler/profilers.py [hunk 2]
```diff
@@ -136,9 +181,16 @@ def __init__(self, output_filename: str = None, line_count_restriction: float =
                 or a decimal fraction between 0.0 and 1.0 inclusive (to select a percentage of lines)
         """
         self.profiled_actions = {}
-        self.output_filename = output_filename
         self.line_count_restriction = line_count_restriction
 
+        self.output_fname = output_filename
+        self.output_file = open(self.output_fname, 'w') if self.output_fname else None
+
+        streaming_out = [log.info]
```
Comment (jeremyjordan) [inline pytorch_lightning/profiler/profilers.py]: i'm not sure we want to always log this report, as it is quite verbose.

Comment (Borda) [reply pytorch_lightning/profiler/profilers.py]: So add a argument to enable this logging?

**Comment by jeremyjordan** [comment]
@Borda mind submitting a follow-up PR which makes `log.info` write stream optional for the `AdvancedProfiler`?

**Comment by Borda** [comment]
I changed it as it was so it is file or info... Or you want to remove info for any case?

**Comment by jeremyjordan** [comment]
ah i missed that last round of commits, thanks!

**Pull request merged**